agent/acs/handler/acs_handler.go

```diff
@@ -138,147 +102,128 @@ type sessionResources interface {
-func StartSession(ctx context.Context, args StartSessionArguments) error {
-	backoff := utils.NewSimpleBackoff(connectionBackoffMin, connectionBackoffMax, connectionBackoffJitter, connectionBackoffMultiplier)
-	session := newSessionResources(args)
-	return startSession(ctx, args, backoff, session)
+// sessionState defines state recorder interface for the
+// session established with ACS. It can be used to record and
+// retrieve data shared across multiple connections to ACS
+type sessionState interface {
+	// connectedToACS callback indicates that the client has
+	// connected to ACS
+	connectedToACS()
+	// getSendCredentialsURLParameter retrieves the value for
+	// the 'sendCredentials' URL parameter
+	getSendCredentialsURLParameter() string
 }
 
-func newSessionResources(args StartSessionArguments) sessionResources {
-	return &acsSessionResources{
-		startSessionArguments: args,
-		sendCredentials:       true,
+// NewSession creates a new Session object
+func NewSession(ctx context.Context,
+	acceptInsecureCert bool,
+	config *config.Config,
+	deregisterInstanceEventStream *eventstream.EventStream,
+	containerInstanceArn string,
+	credentialsProvider *credentials.Credentials,
+	ecsClient api.ECSClient,
+	stateManager statemanager.StateManager,
+	taskEngine engine.TaskEngine,
+	credentialsManager rolecredentials.Manager) *Session {
+	resources := newSessionResources(config.AWSRegion, credentialsProvider, acceptInsecureCert)
+	backoff := utils.NewSimpleBackoff(connectionBackoffMin, connectionBackoffMax,
+		connectionBackoffJitter, connectionBackoffMultiplier)
+	derivedContext, cancel := context.WithCancel(ctx)
+
+	return &Session{
+		acceptInsecureCert:            acceptInsecureCert,
+		agentConfig:                   config,
+		deregisterInstanceEventStream: deregisterInstanceEventStream,
+		containerInstanceARN:          containerInstanceArn,
+		credentialsProvider:           credentialsProvider,
+		ecsClient:                     ecsClient,
+		stateManager:                  stateManager,
+		taskEngine:                    taskEngine,
+		credentialsManager:            credentialsManager,
+		ctx:                           derivedContext,
+		cancel:                        cancel,
+		backoff:                       backoff,
+		resources:                     resources,
 	}
 }
 
-// startSession creates a session with ACS and handles requests from ACS
-// It also tries to repeatedly connect to ACS when disconnected
-func startSession(ctx context.Context, args StartSessionArguments, backoff *utils.SimpleBackoff, acsResources sessionResources) error {
+// Start starts the session. It'll forever keep trying to connect to ACS unless:
+// 1. The context is cancelled
+// 2. Instance is deregistered
+//
+// If the context is cancelled, Start() would return with the error code returned
+// by the context.
+// If the instance is deregistered, Start() would emit an event to the
+// deregister-instance event stream and blocks until the ECS Agent is killed.
```

Comment:
if we kill the agent, tasks/containers that aren't stopped yet on the instance (because the Agent  never receives the message from ACS to kill them) would fail to run if they depend on Agent for their credentials. Also, if `ecs-init` restarts the connection, there's a possibility of the connection storm as well.